### PR TITLE
device: refactor to simplify maintenance

### PR DESCRIPTION
**Pull request by pabigot**
`DEVICE_AND_API_INIT` and `DEVICE_DEFINE` are identical except that `DEVICE_DEFINE` adds a parameter providing the device pm control function, while `DEVICE_AND_API_INIT` does not.  This requires duplicate implementations where if `CONFIG_DEVICE_POWER_MANAGEMENT` is enabled then `DEVICE_AND_API_INIT` delegates to `DEVICE_DEFINE` with a dummy pm control function, and if it is not enabled then `DEVICE_DEFINE` discards the parameter and delegates to `DEVICE_AND_API_INIT`.

`DEVICE_INIT` is like `DEVICE_AND_API_INIT` but doesn't provide an API.

Clean this up by refactoring so that `DEVICE_DEFINE` is the core implementation, providing with and without device power management right next to each other where they can be compared and maintained. Redefine `DEVICE_INIT` and `DEVICE_AND_API_INIT` delegate to `DEVICE_DEFINE`.

Also remove duplicate code by extracting the variations due to enabling device power management into macros.

This is a step intended to simplify providing dependency information for #22545 and related PRs.

### CODE REVIEW

**Review by mbolivar-nordic — COMMENTED** [review]
Looks nice. I find the current organization confusing. One question and a couple of nitpicks.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by carlescufi** [comment]
@tbursztyka could you please take a quick look?

File: include/init.h [hunk 1]
```diff
@@ -70,6 +70,31 @@ void z_sys_init_run_level(int32_t level);
  *
```
Comment (tbursztyka) [inline include/init.h:70]: So Z_* as a prefix to a macro prevents its documentation to be included?

That said, it looks like SYS_INIT() documentation is not present either...

Comment (pabigot) [reply include/init.h:70]: It's added by the changes in this PR.

Comment (tbursztyka) [reply include/init.h:70]: ok no other comments then

**Review by tbursztyka — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**